### DISCUSSION TITLE
Validate xml against YANG model: Invalid identityref error

### DISCUSSION THREAD

**Discussion by smark28** [CESNET/libyang#2176] [Q&A] [answered]
Hello Team, 
I have been trying to find an answer to this question, but without success. 
Asking for help with understanding, what is wrong,  I'm getting a validation error against open-config models.
What am I missing?  Thanks in advance.
 
yanglint openconfig-platform.yang hardware_small.xml
 ```
libyang err : Invalid identityref "ACTIVE" value - identity not found in module "openconfig-platform". (Data location "/openconfig-platform:components/component[name='Rack 0']/state/oper-status", line number 5.)
YANGLINT[E]: Failed to parse input data file "hardware_small.xml".
 ```
 
 hardware_small.xml
 ```
<components xmlns="http://openconfig.net/yang/platform">
   <component>
    <name>Rack 0</name>
    <state>
      <oper-status>ACTIVE</oper-status>
    </state>
   </component>
</components>
 ```
 
openconfig-platform.yang
```
module openconfig-platform {
  yang-version "1";
  namespace "http://openconfig.net/yang/platform";
  prefix "oc-platform";
  import openconfig-platform-types { prefix oc-platform-types; }

  grouping platform-component-state {
    leaf id { type string; }

    leaf oper-status {
      type identityref {
        base oc-platform-types:COMPONENT_OPER_STATUS;
      }
    }
  }

  grouping platform-component-top {
    container components {
      list component {
        key "name";

        leaf name {
          type string;
        }

        container state {
          config false;
          uses platform-component-state;
        }

      }
    }
  }
  uses platform-component-top;
}
```
 
 openconfig-platform-types.yang
```
module openconfig-platform-types {
  yang-version "1";
  namespace "http://openconfig.net/yang/platform-types";
  prefix "oc-platform-types";

  identity COMPONENT_OPER_STATUS {
    description
      "Current operational status of a platform component";
  }

  identity ACTIVE {
    base COMPONENT_OPER_STATUS;
  }
}
```

yanglint -v
yanglint 2.1.157

**Comment by michalvasko** [comment]
I would hope the message is clear
```
Invalid identityref "ACTIVE" value - identity not found in module "openconfig-platform".
```
Because the identity `ACTIVE` is not in `openconfig-platform` but in `openconfig-platform-types`. Try validating
```xml
<components xmlns="http://openconfig.net/yang/platform">
  <component>
   <name>Rack 0</name>
   <state>
     <oper-status xmlns:oct="http://openconfig.net/yang/platform-types">oct:ACTIVE</oper-status>
   </state>
  </component>
</components>
```

**Reply by smark28** [reply]
Hi @michalvasko  

Thanks for for the quick response.
It works, when I additionally implement openconfig-platform-types.yang

Does this mean that it is not enough to specify this identityref in the model, the validation process won't look into related files automatically?
```
    leaf oper-status {
      type identityref {
        base oc-platform-types:COMPONENT_OPER_STATUS;
      }
    }
```
The fact is that I do not prepare the original XML file manually, it is generated by an RPC request from the device. 
If I understand correctly I can't check them against the yang model without manually correcting elements/namespaces, right?

**Comment by michalvasko** [comment, marked as answer]
> Does this mean that it is not enough to specify this identityref in the model, the validation process won't look into related files automatically?

Even according to YANG, you have to specify both the module and the identity name. If for no other reason, then because there may be name collisions.

> If I understand correctly I can't check them against the yang model without manually correcting elements/namespaces, right?

I guess, the generated RPC request is simply invalid.
